README.md
# haveibeenpwned

API methods for Have I been pwned (unofficial)

[![npm](https://img.shields.io/npm/v/haveibeenpwned.svg?maxAge=3600)](https://github.com/fvdm/nodejs-haveibeenpwned/blob/master/CHANGELOG.md)
[![Build Status](https://travis-ci.org/fvdm/nodejs-haveibeenpwned.svg?branch=master)](https://travis-ci.org/fvdm/nodejs-haveibeenpwned)
[![Coverage Status](https://coveralls.io/repos/github/fvdm/nodejs-haveibeenpwned/badge.svg?branch=master)](https://coveralls.io/github/fvdm/nodejs-haveibeenpwned?branch=master)
[![bitHound Dependencies](https://www.bithound.io/github/fvdm/nodejs-haveibeenpwned/badges/master/dependencies.svg)](https://www.bithound.io/github/fvdm/nodejs-haveibeenpwned/master/dependencies/npm)
[![bitHound Code](https://www.bithound.io/github/fvdm/nodejs-haveibeenpwned/badges/master/code.svg)](https://www.bithound.io/github/fvdm/nodejs-haveibeenpwned/master/files)
[![Greenkeeper badge](https://badges.greenkeeper.io/fvdm/nodejs-haveibeenpwned.svg)](https://greenkeeper.io/)


* [Node.js](https://nodejs.org)
* [Have I been pwned](https://haveibeenpwned.com)
* [API documentation](https://haveibeenpwned.com/API/v2)


# Example

```js
const hibp = require ('haveibeenpwned') ();

hibp.breachedAccount ('foo@bar.com', (err, data) => {
  if (err) {
    return console.log (err);
  }

  console.log (data);
});
```


# Installation

`npm i haveibeenpwned`


# Configuration

name    | type   | required | default | description
:-------|:-------|:---------|:--------|:--------------------------
timeout | number | no       | 5000    | Request wait timeout in ms


```js
const pwned = require ('haveibeenpwned') ({
  timeout: 10000
});
```


# Have I Been Pwned

Below are the methods for the main Have I Been Pwned API.

For simplicity no error handling is included in the callback examples.
It's straightforward: when there is a problem `err` is an instance of
_Error_ and `data` is _undefined_, otherwise `err` is _null_ and
`data` is set to the result.


## .breachedAccount

**( account, [params], callback )**

All breaches for an account.


argument | type     | required | description
:--------|:---------|:---------|:--------------------------------
account  | string   | yes      | Email or username to check
params   | object   | no       | Additional parameters to include
callback | function | yes      | `function (err, data) {}`


```js
hibp.breachedAccount ('foo@bar.com', yourCallback);
```

[Example output](https://haveibeenpwned.com/api/v2/breachedaccount/foo@bar.com)


## .breaches

**( [params], callback )**

All breaches in the system.


argument | type     | required | description
:--------|:---------|:---------|:--------------------------------
params   | object   | no       | Additional parameters to include
callback | function | yes      | `function (err, data) {}`


```js
hibp.breaches (yourCallback);
```

[Example output](https://haveibeenpwned.com/api/v2/breaches)


## .breach

**( name, [params], callback )**

A single breached site.


argument | type     | required | description
:--------|:---------|:---------|:-----------------------------------
name     | string   | yes      | Site name to check, i.e. `linkedin`
params   | object   | no       | Additional parameters to include
callback | function | yes      | `function (err, data) {}`


```js
hibp.breach ('Adobe', yourCallback);
```

[Example output](https://haveibeenpwned.com/api/v2/breach/Adobe)


## .pasteAccount

**( account, callback )**

All pastes for an account.


argument | type     | required | description
:--------|:---------|:---------|:--------------------------
account  | string   | yes      | Email or username to check
callback | function | yes      | `function (err, data) {}`


```js
hibp.pasteAccount ('foo@bar.com', yourCallback);
```

[Example output](https://haveibeenpwned.com/api/v2/pasteaccount/foo@bar.com)


## .dataclasses

**( callback )**

All pastes for an account.


argument | type     | required | description
:--------|:---------|:---------|:-------------------------
callback | function | yes      | `function (err, data) {}`


```js
hibp.dataclasses (yourCallback);
```

[Example output](https://haveibeenpwned.com/api/v2/dataclasses)


# Pwned Passwords

Below are the methods for the Pwned Passwords API.


## .pwnedpasswords.byPassword

**( password, [hashed], callback )**

Check if a password was leaked.
The `data` is `0` (int) when none were found.


argument | type     | default | description
:--------|:---------|:--------|:---------------------
password | string   |         | The password to check
hashed   | bool     | false   | Is the password already SHA-1 hashed
callback | function |         | `(err, data)`



```js
hibp.pwnedpasswords.byPassword ('secret', (err, count) => {
  if (!count) {
    console.log ('Great! Password is not found.');
  } else {
    console.log ('Oops! Password was found ' + count + ' times!');
  }
});

// -> Oops! Password was found 195263 times!
```


## .pwnedpasswords.byRange

**( hash, [sort], callback )**

Get password hashes similar to the first 5 characters of the SHA-1 hash
provided. The callback `data` is an *object* where the keys are the
lowercase hashes and the values are the number of times they were used.
When nothing is found `data` is `0` (int).

A hash longer then 5 characters is truncated before being sent to the API
and can be in uppercase or lowercase.


argument | type     | default | description
:--------|:---------|:--------|:-----------------------
hash     | string   |         | The SHA-1 hash to check
sort     | bool     | false   | Sort by counts in descending order
callback | function |         | `(err, data)`


### Sorted by key (default)

```js
hibp.pwnedpasswords.byRange ('abcdef', (err, data) => {
  for (sha in data) {
    console.log ('%s  %i x', sha, data[sha]);
  }
});
```


### Sorted by use counts

```js
hibp.pwnedpasswords.byRange ('abcdef', true, (err, data) => {
  const count = Object.keys (data).length;
  let i = 1;

  console.log ('Found %i matches', count);
  console.log ('Top 3:');

  for (sha in data) {
    console.log ('%i  %s  %i x', i, sha, data[sha]);

    // stop after 3
    if (i === 3) { break; }
    i++;
  }
});

// ->
// Found 511 matches
// Top 3:
// 1  3b8a55c2b3bf42b83e41f0f95a4149043f6  336 x
// 2  6a7fc410810db77855d9dfe5b94b95196f7  304 x
// 3  73a1a21a7b13b50536df19fd586abdf3145  193 x
```


# Unicense

This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <http://unlicense.org>


# Author

[Franklin van de Meent](https://frankl.in)

[![Buy me a coffee](https://frankl.in/u/kofi/kofi-readme.png)](https://ko-fi.com/franklin)


CHANGELOG.md
### 1.1.0 (2018-02-23)

##### Chores

* **package:**  update dotest to version 2.3.0 ([c098f8bc](https://github.com/fvdm/nodejs-haveibeenpwned/commit/c098f8bc4ec4d948a1f48ac0d465cc62f283a310))

##### Documentation Changes

* **readme:**
  *  Add Pwned Passwords methods ([793d2f71](https://github.com/fvdm/nodejs-haveibeenpwned/commit/793d2f71a306b8a9ec1a6b499fb82937d98807a6))
  *  Minor clean up ([260a7819](https://github.com/fvdm/nodejs-haveibeenpwned/commit/260a78190a5cf01cd983e1c44af0aa5fc19656eb))

##### New Features

* **byRange:**  Add optional sorting ([49ee72d7](https://github.com/fvdm/nodejs-haveibeenpwned/commit/49ee72d70e3cfd4b1764caef192d0e7e4de73225))
* **httpRequestPP:**  Pwned Passwords methods ([ec1020c0](https://github.com/fvdm/nodejs-haveibeenpwned/commit/ec1020c0c7cd3be590b0a65673aac04612f22c7b))

##### Refactors

* **export:**  Cleaner timeout setting ([58f17aa9](https://github.com/fvdm/nodejs-haveibeenpwned/commit/58f17aa9397bce12fb33220c212f9a39544d7cf8))
* **httpRequest:**  Renamed to httpRequestHIBP ([418f5eb5](https://github.com/fvdm/nodejs-haveibeenpwned/commit/418f5eb56d1c523899bea970ba5f4ea65507f4f0))

##### Code Style Changes

* **syntax:**  Clean up and modernization ([d70ef7f0](https://github.com/fvdm/nodejs-haveibeenpwned/commit/d70ef7f04cb10e038446382696de15ccdf5269f6))
* **readme:**  Better parsable syntax ([5803e5d6](https://github.com/fvdm/nodejs-haveibeenpwned/commit/5803e5d6358c7fdf79db5d17e80de3588efec5be))

##### Tests

* **main:**
  *  Bypass Node v6.13.0 missing Object.values ([ef3f7766](https://github.com/fvdm/nodejs-haveibeenpwned/commit/ef3f776660b1a328c1d66160caea719c65541fdd))
  *  Method tests for pwnedpasswords ([f05ccc4c](https://github.com/fvdm/nodejs-haveibeenpwned/commit/f05ccc4c72ca5ad8bdbc052b0593fad87c9f3d39))
  *  Interface checks for pwnedpasswords ([44273e53](https://github.com/fvdm/nodejs-haveibeenpwned/commit/44273e534053ecd59ab4f250d7362fe643ee292c))
  *  Add SHA-1 hash function ([f85570c3](https://github.com/fvdm/nodejs-haveibeenpwned/commit/f85570c384df5762b9d43b99c11531d1587ef365))

#### 1.0.3 (2017-12-12)

##### Chores

* **package:**
  * Reduce dev deps to dotest ([fa9e1b13](https://github.com/fvdm/nodejs-haveibeenpwned/commit/fa9e1b1343c7d40d20975fada3ba09139bd9a918))
  * Update dependencies ([e6bd733e](https://github.com/fvdm/nodejs-haveibeenpwned/commit/e6bd733eed52b014a9f4d43bb334b077ea3c47d9))
  * Update dotest dev dep ([935f36a8](https://github.com/fvdm/nodejs-haveibeenpwned/commit/935f36a882e03f17036720188755b31ba3839872))
  * Update dev deps ([69875d00](https://github.com/fvdm/nodejs-haveibeenpwned/commit/69875d00d42e8d1a0a726c381d6dfe8701817bf0))
  * Replaced test runner and dev deps by dotest ([e14f8679](https://github.com/fvdm/nodejs-haveibeenpwned/commit/e14f86795d4986ee428989cb1a8b84b66ec58bd5))
  * update eslint to version 3.0.0 ([36e2933a](https://github.com/fvdm/nodejs-haveibeenpwned/commit/36e2933a0e3b392ca612ebb03cb677f780c25dc9))
* **develop:**
  * Cleanup .gitignore ([70ca7760](https://github.com/fvdm/nodejs-haveibeenpwned/commit/70ca7760545d51c8f12a47ce84386b45fdc42b6a))
  * Added bitHound config ([6f95c3e2](https://github.com/fvdm/nodejs-haveibeenpwned/commit/6f95c3e23a9e15cd3efc3f6b4c2902b6c418fbea))

##### Documentation Changes

* **readme:**
  * Minor edits ([a995fec4](https://github.com/fvdm/nodejs-haveibeenpwned/commit/a995fec4ebe8ed81da6dfaa9cfb97ca07ec52f26))
  * add Greenkeeper badge ([4bedee12](https://github.com/fvdm/nodejs-haveibeenpwned/commit/4bedee124a1e236444f1f22df1a5c12e54fac8f3))
  * Add coffee link to Author ([253de0e9](https://github.com/fvdm/nodejs-haveibeenpwned/commit/253de0e9490d26ef7ea7fc2366a05dbc6dca5f43))
* **badges:**
  * Change bitHound to master ([b58a8cc4](https://github.com/fvdm/nodejs-haveibeenpwned/commit/b58a8cc44528bca752f452601e86db7136812307))
  * Move Greenkeeper to badges ([44a4661a](https://github.com/fvdm/nodejs-haveibeenpwned/commit/44a4661a2b7fc42f3a8afb2a27785511c65fefb1))
  * Added bitHound code quality ([d6a4f400](https://github.com/fvdm/nodejs-haveibeenpwned/commit/d6a4f400b682a16742230a3ad2e5147bb4dd4e5b))

##### Refactors

* **package:** Minimum supported node v4.0 ([6db83d62](https://github.com/fvdm/nodejs-haveibeenpwned/commit/6db83d62dbc831f783719acdc61fc91c5ac4cf36))

##### Code Style Changes

* **comment:** Fix JSdoc syntax ([64b1d22d](https://github.com/fvdm/nodejs-haveibeenpwned/commit/64b1d22dda27222caf2c6db2405891623713c77d))
* **example:** ES6 syntax ([a81ba827](https://github.com/fvdm/nodejs-haveibeenpwned/commit/a81ba82715e28d25134c86a7a2770e1a4be9420d))

##### Tests

* **main:**
  * Fix syntax typos ([3f9edd68](https://github.com/fvdm/nodejs-haveibeenpwned/commit/3f9edd68c1fe423ad581becb1b24e69e92c050e1))
  * Pause 2 seconds between tests ([1df40fa6](https://github.com/fvdm/nodejs-haveibeenpwned/commit/1df40fa6fbba5f0231e07e6a423860740f2c5a89))
  * Replaced invalid check example.net ([9619424f](https://github.com/fvdm/nodejs-haveibeenpwned/commit/9619424fdd7fb1a0310fae8c7d5e2d3ad590c992))
* **config:**
  * Allow 500 lines in .bithoundrc ([ac3f7f99](https://github.com/fvdm/nodejs-haveibeenpwned/commit/ac3f7f99e47726705dc935f980ca10b01aeab11d))
  * Remove ecmaVersion from .eslintrc ([72840934](https://github.com/fvdm/nodejs-haveibeenpwned/commit/72840934b807a960e043d3db95a1a18102e34605))
  * Update Travis CI node versions ([770ab612](https://github.com/fvdm/nodejs-haveibeenpwned/commit/770ab612c911e529287f85692378688b393609a7))
  * Use dynamic node versions on Travis CI ([4fce2800](https://github.com/fvdm/nodejs-haveibeenpwned/commit/4fce2800772b6d941d62794f3fd8a5fa8e73390c))
* **style:**
  * Convert to ES6 syntax ([147f862c](https://github.com/fvdm/nodejs-haveibeenpwned/commit/147f862c3199d8c65a39c905b8235ed6b0adacf7))
  * Cleanup intro comment ([06ee70d7](https://github.com/fvdm/nodejs-haveibeenpwned/commit/06ee70d76f723e297dac46c7ca21d00212416f59))

test.js
/*
Name:         haveibeenpwned - test.js
Description:  Unit tests for index.js
Author:       Franklin van de Meent (https://frankl.in)
Source code:  https://github.com/fvdm/nodejs-haveibeenpwned
Feedback:     https://github.com/fvdm/nodejs-haveibeenpwned/issues
License:      Unlicense (public domain, see LICENSE file)
*/


const dotest = require ('dotest');
const crypto = require ('crypto');
const app = require ('./');

const config = {
  timeout: process.env.Timeout || null
};

const pwned = app (config);

const testPass = 'doesnt-exist,' + Date.now();


/**
 * Generate SHA-1 hash from a string
 *
 * @return  {string}       SHA-1 hash as hex string
 * @param   {mixed}   str  Data to hash
 */

function hashSha1 (str) {
  return crypto
    .createHash ('sha1')
    .update (str)
    .digest ('hex');
}


dotest.add ('Interface', test => {
  const pp = pwned && pwned.pwnedpasswords;

  test()
    .isFunction ('fail', 'exports', app)
    .isObject ('fail', 'interface', pwned)
    .isObject ('fail', '.pwnedpasswords', pp);

  if (pwned) {
    test()
      .isFunction ('fail', '.breachedAccount', pwned.breachedAccount)
      .isFunction ('fail', '.breaches', pwned.breaches)
      .isFunction ('fail', '.breach', pwned.breach)
      .isFunction ('fail', '.pasteAccount', pwned.pasteAccount)
      .isFunction ('fail', '.dataclasses', pwned.dataclasses);
  }

  if (pp) {
    test()
      .isFunction ('fail', '.pwnedpasswords.byPassword', pp && pp.byPassword)
      .isFunction ('fail', '.pwnedpasswords.byRange', pp && pp.byRange);
  }


  test().done();
});


dotest.add ('Method .breachedAccount - without params', test => {
  pwned.breachedAccount ('foo@bar.com', (err, data) => {
    const item = data && data [0];

    test (err)
      .isArray ('fail', 'data', data)
      .isNotEmpty ('fail', 'data', data)
      .isObject ('fail', 'data[0]', item)
      .isNotEmpty ('fail', 'data[0]', item)
      .done();
  });
});


dotest.add ('Method .breachedAccount - with params', test => {
  const params = {
    domain: 'acne.org'
  };

  pwned.breachedAccount ('foo@bar.com', params, (err, data) => {
    const item = data && data [0];

    test (err)
      .isArray ('fail', 'data', data)
      .isNotEmpty ('fail', 'data', data)
      .isObject ('fail', 'data[0]', item)
      .isNotEmpty ('fail', 'data[0]', item)
      .done();
  });
});


dotest.add ('Method .breaches - without params', test => {
  pwned.breaches ((err, data) => {
    const item = data && data [0];

    test (err)
      .isArray ('fail', 'data', data)
      .isNotEmpty ('fail', 'data', data)
      .isObject ('fail', 'data[0]', item)
      .isNotEmpty ('fail', 'data[0]', item)
      .done();
  });
});


dotest.add ('Method .breaches - with params', test => {
  const params = {
    domain: 'linkedin.com'
  };

  pwned.breaches (params, (err, data) => {
    const item = data && data [0];

    test (err)
      .isArray ('fail', 'data', data)
      .isNotEmpty ('fail', 'data', data)
      .isObject ('fail', 'data[0]', item)
      .isNotEmpty ('fail', 'data[0]', item)
      .done();
  });
});


dotest.add ('Method .breach', test => {
  pwned.breach ('LinkedIn', (err, data) => {
    test (err)
      .isObject ('fail', 'data', data)
      .isNotEmpty ('fail', 'data', data)
      .done();
  });
});


dotest.add ('Method .pasteAccount', test => {
  pwned.pasteAccount ('foo@bar.com', (err, data) => {
    const item = data && data [0];

    test (err)
      .isArray ('fail', 'data', data)
      .isNotEmpty ('fail', 'data', data)
      .isObject ('fail', 'data[0]', item)
      .isNotEmpty ('fail', 'data[0]', item)
      .done();
  });
});


dotest.add ('Method .dataclasses', test => {
  pwned.dataclasses ((err, data) => {
    const item = data && data [0];

    test (err)
      .isArray ('fail', 'data', data)
      .isString ('fail', 'data[0]', item)
      .isNotEmpty ('fail', 'data[0]', item)
      .done();
  });
});


dotest.add ('Method .pwnedpassword.byPassword - plain, none found', test => {
  pwned.pwnedpasswords.byPassword (testPass, (err, data) => {
    test (err)
      .info ('Password: ' + testPass)
      .info (' ')
      .isNull ('fail', 'err', err)
      .isExactly ('fail', 'data', data, 0)
      .done();
  });
});


dotest.add ('Method .pwnedpasswords.byPassword - hashed, none found', test => {
  const hash = hashSha1 (testPass);

  pwned.pwnedpasswords.byPassword (hash, true, (err, data) => {
    test (err)
      .info ('Password: ' + testPass)
      .info ('Hash:     ' + hash)
      .info (' ')
      .isNull ('fail', 'err', err)
      .isExactly ('fail', 'data', data, 0)
      .done();
  });
});


dotest.add ('Method .pwnedpasswords.byPassword - plain, results', test => {
  pwned.pwnedpasswords.byPassword ('12345', (err, data) => {
    test (err)
      .info ('Password: 12345')
      .info (' ')
      .isNull ('fail', 'err', err)
      .isNumber ('fail', 'data', data)
      .isCondition ('fail', 'data', data, '>', 0)
      .done();
  });
});


dotest.add ('Method .pwnedpasswords.byRange - results', test => {
  const hash = hashSha1 ('12345');

  pwned.pwnedpasswords.byRange (hash, (err, data) => {
    const item = data && data['37d0679ca88db6464eac60da96345513964'];

    test (err)
      .info ('Password: 12345')
      .info ('Hash:     ' + hash)
      .info ('Item:     37d0679ca88db6464eac60da96345513964')
      .info (' ')
      .isNull ('fail', 'err', err)
      .isObject ('fail', 'data', data)
      .isNumber ('fail', 'data[item]', item)
      .isCondition ('fail', 'data[item]', item, '>', 0)
      .done();
  });
});


dotest.add ('Method .pwnedpasswords.byRange - sorted', test => {
  const hash = hashSha1 ('12345');

  pwned.pwnedpasswords.byRange (hash, true, (err, data) => {
    const item = data && data['37d0679ca88db6464eac60da96345513964'];
    const values = Object.keys (data).map (key => {
      return data [key];
    });
    const first = values.shift();
    const second = values.shift();
    const third = values.shift();

    test (err)
      .info ('Password: 12345')
      .info ('Hash:     ' + hash)
      .info ('Item:     37d0679ca88db6464eac60da96345513964')
      .info (' ')
      .isNull ('fail', 'err', err)
      .isObject ('fail', 'data', data)
      .isNumber ('fail', 'data[item]', item)
      .isCondition ('fail', 'data[item]', item, '>', 0)
      .isCondition ('fail', 'order 1 vs 2', first, '>', second)
      .isCondition ('fail', 'order 2 vs 3', second, '>', third)
      .done();
  });
});


dotest.add ('Error: invalid hash prefix', test => {
  pwned.pwnedpasswords.byRange ('test', (err, data) => {
    test ()
      .isError ('fail', 'err', err)
      .isExactly ('warn', 'err.message', err && err.message, 'The hash prefix was not in a valid format')
      .isUndefined ('fail', 'data', data)
      .done();
  });
});


dotest.add ('Error: not found', test => {
  pwned.breachedAccount ('info@invalid--example.net', (err, data) => {
    test()
      .isError ('fail', 'err', err)
      .isExactly ('fail', 'err.message', err && err.message, 'not found')
      .isExactly ('fail', 'err.statusCode', err && err.statusCode, 404)
      .isUndefined ('fail', 'data', data)
      .done();
  });
});


dotest.add ('Error: API error', test => {
  pwned.breachedAccount ('', (err, data) => {
    test()
      .isError ('fail', 'err', err)
      .isNumber ('fail', 'err.statusCode', err && err.statusCode)
      .isUndefined ('fail', 'data', data)
      .done();
  });
});


dotest.add ('Error: request timed out - haveibeenpwned API', test => {
  const tmp = app ({
    timeout: 1
  });

  tmp.dataclasses ((err, data) => {
    test()
      .isError ('fail', 'err', err)
      .isExactly ('fail', 'err.code', err && err.code, 'TIMEOUT')
      .isUndefined ('fail', 'data', data)
      .done();
  });
});


dotest.add ('Error: request timed out - pwnedpasswords API', test => {
  const tmp = app ({
    timeout: 1
  });

  tmp.pwnedpasswords.byPassword ('test', (err, data) => {
    test()
      .isError ('fail', 'err', err)
      .isExactly ('fail', 'err.code', err && err.code, 'TIMEOUT')
      .isUndefined ('fail', 'data', data)
      .done();
  });
});


dotest.run (2000);


index.js
/*
Name:         haveibeenpwned - index.js
Description:  API methods for HaveIBeenPwnd.com (unofficial)
Author:       Franklin van de Meent (https://frankl.in)
Source code:  https://github.com/fvdm/nodejs-haveibeenpwned
Feedback:     https://github.com/fvdm/nodejs-haveibeenpwned/issues
License:      Unlicense (public domain, see LICENSE file)
*/


const httpreq = require ('httpreq');

let config = {
  timeout: 5000,
  userAgent: 'nodejs-haveibeenpwned (https://github.com/fvdm/nodejs-haveibeenpwned)'
};


/**
 * Sort an object by its values
 * in descending order
 *
 * @return  {object}       The sorted object
 * @param   {object}  obj  The object to sort
 */

function sortObjectByValues (obj) {
  let sorting = [];
  let result = {};
  let key;

  for (key in obj) {
    sorting.push ({ key, count: obj [key] });
  }

  sorting.sort ((a, b) => {
    if (a.count > b.count) {
      return -1;
    }

    if (a.count < b.count) {
      return 1;
    }

    return 0;
  });

  sorting.forEach (itm => {
    result [itm.key] = itm.count;
  });

  return result;
}


/**
 * Process API error
 *
 * @callback  callback
 * @return    {void}
 *
 * @param     {object}    res       httpreq response details
 * @param     {Error}     err       Additional Error to include
 * @param     {function}  callback  `(err, data)`
 */

function processApiError (res, err, callback) {
  let error = new Error ('API error');

  error.statusCode = res.statusCode;
  error.body = res.body;
  error.error = err;
  callback (error);
}


/**
 * Process HTTP response
 *
 * @callback  callback
 * @return    {void}
 *
 * @param     {Error|null}  err       httpreq error
 * @param     {object}      res       httpreq response details
 * @param     {function}    callback  `(err, data)`
 */

function processResponse (err, res, callback) {
  let data;
  let error;

  if (err) {
    return callback (err);
  }

  if (res.statusCode === 404) {
    error = new Error ('not found');
    error.statusCode = res.statusCode;
    return callback (error);
  }

  try {
    data = JSON.parse (res.body);
  } catch (e) {
    return processApiError (res, e, callback);
  }

  return callback (null, data);
}


/**
 * Send HTTP request to API
 *
 * @callback  callback
 * @return    {void}
 *
 * @param     {string}     service   API service name
 * @param     {string}     method    API service method name
 * @param     {object}     [params]  Additional parameters to include
 * @param     {function}   callback  `(err, data)`
 */

function httpRequestHIBP (service, method, params, callback) {
  let options = {
    url: 'https://haveibeenpwned.com/api/v2/' + service + '/' + method,
    method: 'GET',
    timeout: config.timeout,
    headers: {
      'User-Agent': config.userAgent
    }
  };

  if (typeof params === 'function') {
    callback = params;
    params = {};
  }

  options.parameters = params;

  httpreq.doRequest (options, (err, res) => {
    processResponse (err, res, callback);
  });
}


/**
 * All breaches for an account
 *
 * @callback  callback
 * @return    {void}
 *
 * @param     {string}    account   Email or username
 * @param     {object}    [params]  Additional parameters to include
 * @param     {function}  callback  `(err, data)`
 */

function methodBreachedAccount (account, params, callback) {
  httpRequestHIBP ('breachedaccount', account, params, callback);
}


/**
 * All breached sites in the system
 *
 * @callback  callback
 ^ @return    {void}
 *
 * @param     {object}    [params]  Additional parameters to include
 * @param     {function}  callback  `(err, data)`
 */

function methodBreaches (params, callback) {
  httpRequestHIBP ('breaches', '', params, callback);
}


/**
 * A single breached site
 *
 * @callback  callback
 * @return    {void}
 *
 * @param     {string}    name      Email or username
 * @param     {function}  callback  `(err, data)`
 */

function methodBreach (name, callback) {
  httpRequestHIBP ('breach', name, callback);
}


/**
 * All pastes for an account
 *
 * @callback  callback
 * @return    {void}
 *
 * @param     {string}    amount    Email or username
 * @param     {function}  callback  `(err, data)`
 */

function methodPasteAccount (account, callback) {
  httpRequestHIBP ('pasteaccount', account, callback);
}


/**
 * All data classes in the system
 *
 * @callback  callback
 * @return    {void}
 *
 * @param     {function}  callback  `(err, data)`
 */

function methodDataclasses (callback) {
  httpRequestHIBP ('dataclasses', '', callback);
}


/**
 * Send HTTP request to API
 *
 * @callback  callback
 * @return    {void}
 *
 * @param     {string}     path      API method path
 * @param     {object}     [params]  Additional parameters to include
 * @param     {function}   callback  `(err, data)`
 */

function httpRequestPP (path, params, callback) {
  let options = {
    url: 'https://api.pwnedpasswords.com/' + path,
    method: 'GET',
    timeout: config.timeout,
    headers: {
      'User-Agent': config.userAgent
    }
  };

  if (typeof params === 'function') {
    callback = params;
    params = {};
  }

  options.parameters = params;

  httpreq.doRequest (options, (err, res) => {
    let error;
    let msg;

    if (err) {
      callback (err);
    } else if (res.statusCode === 404) {
      callback (null, 0);
    } else if (res.statusCode >= 300) {
      msg = res.body
        .replace (/.+<p>(.+)<\/p>.+/, '$')
        .trim();
      error = new Error (msg);
      error.statusCode = res.statusCode;
      error.body = res.body;

      callback (error);
    } else {
      callback (null, res.body);
    }
  });
}


/**
 * Search the Pwned Passwords database for a password
 *
 * Password not found: callback `data` is {int} `0`
 * Password found:     callback `data` is {int} amount
 *
 * @callback  callback
 * @return    {void}
 *
 * @param     {string}    password  Password to check
 * @param     {bool}      [hashed]  Is the password already SHA-1 hashed
 * @param     {function}  callback  `(err, data)`
 */

function methodPwnedPasswordsByPassword (password, hashed, callback) {
  let params = {};

  if (typeof hashed === 'function') {
    callback = hashed;
    hashed = false;
  }

  if (hashed) {
    params.originalPasswordIsAHash = 'true';
  }

  httpRequestPP ('pwnedpassword/' + password, params, (err, data) => {
    if (err) {
      callback (err);
    } else {
      callback (null, parseInt (data, 10));
    }
  });
}


/**
 * Search the Pwned Passwords database for the first 5 chars of a hash
 * and callback the matched hashed with their counts.
 *
 * Not found: callback `data` is {int} `0`
 * Found:     callback `data` is {object}
 *
 * @callback  callback
 * @return    {void}
 *
 * @param     {string}    hash      SHA-1 hash to check
 * @param     {bool}      [sort]    Sort by counts in descending order
 * @param     {function}  callback  `(err, data)`
 */

function methodPwnedPasswordsByRange (hash, sort, callback) {
  hash = hash.substr (0, 5);

  if (typeof sort === 'function') {
    callback = sort;
    sort = false;
  }

  httpRequestPP ('range/' + hash, (err, data) => {
    let result = {};
    let i;
    let str;
    let sha;

    if (err) {
      callback (err);
      return;
    }

    data = data.trim().split ('\n');

    if (data.length) {
      for (i = 0; i < data.length; i++) {
        str = data[i].split (':');
        sha = str[0].toLowerCase();
        result[sha] = parseInt (str[1], 10);
      }

      if (sort) {
        result = sortObjectByValues (result);
      }

      callback (null, result);
      return;
    }

    callback (null, 0);
  });
}


/**
 * Module interface
 *
 * @return  {object}                      Methods
 *
 * @param   {object}  set                 Configuration params
 * @param   {int}     [set.timeout=5000]  Wait timeout in ms
 */

module.exports = set => {
  if (set && set.timeout) {
    config.timeout = set.timeout;
  }

  return {
    breachedAccount: methodBreachedAccount,
    breaches: methodBreaches,
    breach: methodBreach,
    pasteAccount: methodPasteAccount,
    dataclasses: methodDataclasses,
    pwnedpasswords: {
      byPassword: methodPwnedPasswordsByPassword,
      byRange: methodPwnedPasswordsByRange
    }
  };
};
